### PR TITLE
iOS - Testing for Sensitive Data in Memory

### PR DESCRIPTION
**Pull request by pmilosev**
Opening a PR earlier for review.
I'm still working on the Dynamic Analysis and the Remediation sections.
More commits will follow.

#620

### CODE REVIEW

**Comment by pmilosev** [comment]
I ran out of proper words when describing string mutability in Swift.
I don't fill I have expressed myself well there, but it's the best I could do at the moment :)
There are things like "XYZ can be changed but not modified", or suggestions that something is not mutable but you can change it's value.
Feel free to suggest some better wording there.

File: Document/0x06d-Testing-Data-Storage.md [hunk 1]
```diff
@@ -761,16 +761,53 @@ This will cause the background image to be set to the "overlayImage.png" instead
...
+
+_Swift_ data types, other than collections should strictly be avoided, regardless of whether they are considered mutable or not. Many data types in _Swift_ hold their data by value, not reference. While for simple types like `char` or `int` this allows us to modify their actual memory, having a complex type like `String` handled by value implies a hidden layer of objects, structures, or primitive arrays whose memory can not be directly accessed and modified. Certain usage may appear, and even be documented, to result in mutable data object, but it actually results in a mutable identifier (variable) as opposite to an immutable identifier (constant). For example, many consider the following to result in a mutable `String` in _Swift_, but actually it is an example of a variable whose complex value can be changed (replaced, not modified in place):
+
+```swift
+var str1 = "Goodby"              // "Goodby", base adress:             0x0001039e8dd0
```
Comment (Sjord) [inline Document/0x06d-Testing-Data-Storage.md]: "adress" should be "address". "Goodby" should be "goodbye".

File: Document/0x06d-Testing-Data-Storage.md [hunk 2]
```diff
@@ -761,16 +761,53 @@ This will cause the background image to be set to the "overlayImage.png" instead
...
+
+As we've seen so far usage of _Swift_ or _Objective-C_ data types require deep understanding of the language implementation itself. Further, we have witnessed some core re-factoring between mayor _Swift_ versions, resulting in incompatible behavior of many data types. In order to avoid these issues we recommend usage of primitive data types whenever certain data needs to be securely erased from memory.
+
+Unfortunately, not many libraries and frameworks are designed to allow overwriting of sensitive data. Even Apple does not consider this issue in the official API of the iOS SDK. For example, most of the APIs for data transformation (passers, serializes, etc.) operate on non-primitive data types. Similarly, regardless of weather you flag certain `UITextField` as _Secure Text Entry_ or not, it always return data as `String` or `NSString`.
```
Comment (Sjord) [inline Document/0x06d-Testing-Data-Storage.md]: > it always return

This should be "it always returns".

File: Document/0x06d-Testing-Data-Storage.md [hunk 3]
```diff
@@ -761,16 +761,53 @@ This will cause the background image to be set to the "overlayImage.png" instead
...
 
+To summarize, when performing static analysis for sensitive data exposed in memory, you should:
+- Try to identify application components and make a map of where certain data is used.
```
Comment (Sjord) [inline Document/0x06d-Testing-Data-Storage.md]: > where certain data is used

I am not sure you can have "certain data", if you consider data an uncountable noun.

File: Document/0x06d-Testing-Data-Storage.md [hunk 4]
```diff
@@ -761,16 +761,53 @@ This will cause the background image to be set to the "overlayImage.png" instead
...
+- Verify that sensitive data is handled in as few components as possible.
+- Verify that object references are properly removed, once the object containing sensitive data is no longer needed.
+- For highly sensitive information, verify data is overwritten as soon as it is no longer needed.
+  - Such data must not be passed over immutable data types such as `String` or `NSString`.
+  - Non-primitive data types are in general indicative and should be avoided.
```
Comment (Sjord) [inline Document/0x06d-Testing-Data-Storage.md]: Indicative of what?

File: Document/0x06d-Testing-Data-Storage.md [hunk 5]
```diff
@@ -761,16 +761,53 @@ This will cause the background image to be set to the "overlayImage.png" instead
...
+  - Overwriting should be done before removing references.
+  - Pay attention to third-party components (libraries and frameworks).
+    Good indicator is their public API. Is the sensitive data passing the public API handled as proposed in this chapter ?
```
Comment (Sjord) [inline Document/0x06d-Testing-Data-Storage.md]: > Good indicator is their public API.

An indicator of what?

> Is the sensitive data passing the public API handled as proposed in this chapter ? 

It took me a couple of reads before I understood this sentence. Maybe you can simplify it a bit?

File: Document/0x06d-Testing-Data-Storage.md (hunk 3 above)
Comment (pmilosev) [inline Document/0x06d-Testing-Data-Storage.md]: would "some particular data" be a better option ?

File: Document/0x06d-Testing-Data-Storage.md (hunk 4 above)
Comment (pmilosev) [inline Document/0x06d-Testing-Data-Storage.md]: Probably not the best choice of words here.
In my native language (Macedonian), we use the term "indicative" when something is suspicious and indicates that there is something strange or wrong about the subject.
So the long version would be "Usage of non-primitive data types suggests that data is not probably being handled according to the discussion above and therefore should be avoided."
But this is too long for a summary bullet point.

**Comment by pmilosev** [comment]
@Sjord thanks for the review.

@others: The PR is still not ready to be merged though - I'm working on the dynamic analysis section.

**Comment by b-mueller** [comment]
@pmilosev is this PR ready to be merged now?

**Comment by pmilosev** [comment]
No, not yet, sorry.
I'm working on the dynamic section.
I would like to extend it with some pros and cons of each tool and add the missing content for Needle.
I would try to finish it by the end of this week.

**Comment by pmilosev** [comment]
Hi @b-mueller,

I won't be able to finish the dynamic section in the next period, sorry.
Therefore I propose that you merge the pull request as it is.
Then I can open another one if no one else covers the topic.

**Pull request merged**